src/Package/Impl/Sql/Publish/SqlPublishOptionsDialogViewModel.cs

```diff
@@ -0,0 +1,315 @@
+                    Targets = new List<string>();
+                    break;
+
+                case PublishTargetType.Project:
+                    PopulateProjectList();
+                    break;
+
+                case PublishTargetType.Database:
+                    InitializationTask = PopulateConnectionsListAsync();
+                    TargetHasName = false;
+                    break;
+            }
+        }
+
+        private Task PopulateConnectionsListAsync() {
+            return GetDatabaseConnectionsAsync((connections) => {
+                var index = -1;
+                _targetConnections = connections;
+                if (!string.IsNullOrEmpty(Settings.TargetDatabaseConnection)) {
+                    var indices = _targetConnections.IndexWhere(c => c.ConnectionString.EqualsIgnoreCase(Settings.TargetDatabaseConnection));
+                    index = indices.DefaultIfEmpty(-1).First();
+                }
+                Targets = _targetConnections.Select(c => c.Name).ToList();
+                ForceSelectedTargetIndexUpdate();
+                SelectedTargetIndex = index >= 0 ? index : 0;
+                UpdateState();
+            });
+        }
+
+        private void PopulateProjectList() {
+            _targetProjects = GetDatabaseProjects();
+            if (_targetProjects.Count > 0) {
+                var index = -1;
+                if (!string.IsNullOrEmpty(Settings.TargetProject)) {
+                    var indices = _targetProjects.IndexWhere(x => x.EqualsIgnoreCase(Settings.TargetProject));
+                    index = indices.DefaultIfEmpty(-1).First();
+                }
+                Targets = _targetProjects;
+                ForceSelectedTargetIndexUpdate();
+                SelectedTargetIndex = index >= 0 ? index : 0;
+            }
+        }
+
+        private void ForceSelectedTargetIndexUpdate() {
+            _selectedTargetIndex = -1;
+        }
+
+        private IReadOnlyList<string> GetDatabaseProjects() {
+            var projectMap = new Dictionary<string, EnvDTE.Project>();
+            var solution = _pss.GetSolution();
+            var projects = new List<string>();
+            foreach (EnvDTE.Project project in solution.Projects) {
+                try {
+                    // Some projects throw 'not implemented'
+                    var projectFileName = project.FileName;
+                    if (!string.IsNullOrEmpty(projectFileName) && Path.GetExtension(projectFileName).EqualsIgnoreCase(".sqlproj")) {
+                        projects.Add(project.Name);
+                    }
+                } catch (NotImplementedException) { } catch (COMException) { }
+            }
+
+            if (projects.Count == 0) {
+                projects.Add(Resources.SqlPublishDialog_NoDatabaseProjects);
+            }
+            return projects;
+        }
+
+        private Task GetDatabaseConnectionsAsync(Action<IReadOnlyList<DbConnectionData>> action) {
```

Comment:
You actually don't even need an `action` callback and `SwitchToMainThreadAsync` - just make this method return `connections` - caller is in UI thread already, so you don't need to switch context.

---
In reply to: [73168059](https://github.com/Microsoft/RTVS/pull/2135#discussion_r73168059) [](ancestors = 73168059)